### PR TITLE
[model, ops, trainer] refactor: unify MoE dispatch through OpSlot

### PR DESCRIPTION
**Pull request by TimYangst**
### What does this PR do?

> Migrates the four remaining v4 MoE models (`qwen3_moe`, `qwen3_vl_moe`,
> `qwen3_omni_moe`, `deepseek_v3`) from the legacy `config._moe_implementation`
> flag onto the `OpSlot("moe_experts", "standard")` pattern that
> `qwen3_5_moe` already uses. With the migration complete, the pre-load
> `apply_moe_patch_transformers_v4` block in `auto.py` is gone —
> fused-MoE binding now happens uniformly via `_bind_veomni_ops` *after* the
> model is constructed, so the kernel selection no longer has to be threaded
> through `build_foundation_model` or nested HF configs.
>
> Two small DX folds that became natural once the dispatch was uniform:
>
> - **`cross_entropy_loss_implementation` defaults to `"chunk_loss"`** —
>   the kernel composes `F.linear` + eager CE in a chunked autograd
>   `Function`, so it is hardware-agnostic and runs on both CUDA and Ascend
>   NPU. The legacy `"npu"` spelling is kept as a back-compat alias for the
>   same kernel that additionally gates on `torch_npu`.
> - **`moe_implementation: "fused"`** is rewritten in
>   `OpsImplementationConfig.__post_init__` to `"fused_quack"` on GPU /
>   `"fused_npu"` on NPU with a deprecation warning, so pre-#678 YAMLs keep
>   working through one upgrade window.

### Checklist Before Starting

- Builds on #678 (kernel registry refactor) and #696 (runtime ckpt converter for v4 MoE).
- PR title follows `[{modules}] {type}: {description}`.

### Test

> - `tests/models/test_models_patch.py` (8 cases on transformers v5,
>   including `qwen3_5_moe`, `qwen3_vl_moe`, `qwen3_omni_moe`): pass.
> - `tests/models/test_moe_v4_runtime_merge.py`: the
>   `Qwen3OmniMoeV4FactoryFusedGate` test class is updated to assert the
>   converter always fires now that the thinker uses stacked params in both
>   modes.
> - `tests/e2e/test_e2e_parallel.py` `qwen3_moe` text-parallel-align (SP/EP
>   grid): pass on both v4 and v5. Materialization is now seed-deterministic
>   on CPU, so the EP=1 vs EP=2 grad-norm comparison is no longer flaky.
> - `tests/ops/` (86 passed) including
>   `test_kernel_registry_numerical.py`, `test_moe_hw_gate.py`,
>   `test_fused_moe_split_vs_merged.py`.
> - `make patchgen` clean (4 generated files refreshed). `make quality` clean.

### API and Usage Example

> The user-facing YAML surface is unchanged.
> `model.ops_implementation.moe_implementation` accepts the same values as
> before; selection now reaches the modeling code via OpSlot instead of
> `config._moe_implementation`.
>
> ```yaml
> model:
>   ops_implementation:
>     moe_implementation: fused_triton    # fused_quack | fused_npu | eager
>     # Pre-#678 YAMLs that still say "fused" keep working — auto-resolved
>     # to fused_quack on GPU / fused_npu on NPU, with a one-time warning.
>     # moe_implementation: fused
>     cross_entropy_loss_implementation: chunk_loss   # NEW default; CUDA + NPU
> ```
>
> In the modeling code (v4 monkey-patched and v5 patchgen alike):
>
> ```python
> veomni_moe_experts_forward = OpSlot("moe_experts", "standard")  # module level
>
> class PatchSomeMoeExperts(nn.Module):
>     def forward(self, h, idx, w):
>         if veomni_moe_experts_forward.use_non_eager_impl:
>             return fused_moe_forward(...)
>         # eager fall-through below
> ```
>
> `_bind_veomni_ops` resolves `OpsImplementationConfig.moe_implementation`
> against `KERNEL_REGISTRY`, binds the slot, and (when non-eager) installs
> the global `_fused_moe_forward` pointer used inside the guard.
>
> **Python-side API notes.** The `moe_implementation=` kwarg on
> `build_foundation_model` is removed — kernel selection flows through
> `OpsImplementationConfig.moe_implementation` instead. All in-tree callers
> (`BaseTrainer`, `VLMTrainer`, `TextDPOTrainer`, `DiTTrainer`,
> `tasks/deprecated_task/train_*.py`) are updated. External code that pokes
> `config._moe_implementation` directly should migrate to
> `OpsImplementationConfig.moe_implementation` + `apply_ops_config(...)`.

### Design & Code Changes

> **Per-model migrations.** Each holdout drops
> `self._moe_implementation = getattr(config, "_moe_implementation", "eager")`
> and replaces `if self._moe_implementation == "fused":` with
> `if veomni_moe_experts_forward.use_non_eager_impl:`. A module-level
> `OpSlot("moe_experts", "standard")` is added so `_bind_veomni_ops` (which
> walks `sys.modules[model.__class__.__module__]`) can find it.
>
> - **qwen3_moe** / **deepseek_v3**: the patch entry point mirrors the
>   OpSlot onto the upstream HF module (`hf_qwen3_moe.veomni_moe_experts_forward
>   = veomni_moe_experts_forward`). Needed because on transformers v4 the
>   resolved `model.__class__` lives in the HF module, not VeOmni's
>   modeling file. Same Python object, just an extra reference.
> - **qwen3_vl_moe**: the `Qwen3VLMoeModel.__init__` override (whose only
>   job was propagating `_moe_implementation` to `text_config`) is dropped —
>   its body was otherwise identical to upstream's. Storage stays on the
>   upstream HF transposed `gate_up_proj [E, H, 2*I]` layout.
> - **qwen3_omni_moe**: dual-storage in `Qwen3OmniMoeThinkerTextSparseMoeBlock`
>   collapsed onto stacked-params `Qwen3OmniMoeThinkerExperts` for both
>   modes; the eager loop is `F.linear` over the stacked tensors. The
>   runtime checkpoint converter (`checkpoint_tensor_converter_v4.py`) now
>   fires unconditionally for thinker keys, so HF per-expert checkpoints
>   continue to load. The EP guard at the old `_moe_implementation == "eager"`
>   branch is migrated to `not slot.use_non_eager_impl` and points users at
>   the user-facing config field in the error message.
>
> **`_bind_veomni_ops` extension** (`veomni/models/auto.py`). When iterating
> slots, if a `moe_experts` OpSlot binds to a non-eager kernel,
> `apply_veomni_fused_moe_patch(fused_moe_kernel=...)` is invoked to set the
> global `_fused_moe_forward` pointer that the inline `fused_moe_forward(...)`
> calls inside the OpSlot guards depend on. Eager bindings leave the
> pointer untouched.
>
> **`auto.py` cleanup.** Removes `apply_moe_patch_transformers_v4` (~30 lines),
> `_module_has_opslot`, the `pre_load_modeling_module` block in
> `build_foundation_model` (~25 lines including a multi-paragraph TODO), the
> now-unused `moe_implementation` kwarg, and the `get_model_class` import
> that only the legacy block needed. Net: ~80 lines of plumbing gone.
>
> **`chunk_loss` as default CE.**
> `OpsImplementationConfig.cross_entropy_loss_implementation` flips from
> `"eager"` to `"chunk_loss"`. `install_loss_mapping("chunk_loss")` binds
> the kernel for `ForCausalLM`/`ForConditionalGeneration`;
> `ForSequenceClassification` stays on the eager wrapper because chunk_loss
> hard-codes the causal label shift. `KERNEL_REGISTRY` registers
> `chunk_loss` for OpSlot dispatch under a new `device_type="any"` gate; an
> eager-CE fallback is registered for the `seq_cls` variant so models like
> qwen3 v5 that declare both slots bind cleanly. The legacy `"npu"`
> spelling resolves to the same kernel.
>
> **`moe_implementation: "fused"` legacy alias.** Rewritten in
> `OpsImplementationConfig.__post_init__` to `"fused_quack"` on GPU /
> `"fused_npu"` on NPU with a one-time deprecation warning. Resolved at
> config-parse time so every downstream reader
> (`apply_veomni_fused_moe_patch`, `_bind_veomni_ops`, kernel-registry
> validators) sees the canonical value.
>
> **Trainer call-site updates.** `BaseTrainer`, `VLMTrainer`, `TextDPOTrainer`,
> `DiTTrainer`, and the two deprecated task scripts in
> `tasks/deprecated_task/` no longer pass `moe_implementation=...` to
> `build_foundation_model`.
>
> **Test determinism.** `tests/e2e/test_e2e_parallel.py::_materialize_weights_dir`
> seeds `torch.manual_seed(0)` and initializes on CPU so the materialized
> checkpoint is bit-identical across pytest invocations and GPU
> architectures (CI's L20 vs developers' A100). Without this, the SP/EP
> grid was flaky at the toy-config scale.
>
> **Docs.** `docs/transformers_v5/transformers_v5_moe_weight_loading.md`,
> `docs/design/kernel_selection.md`, `docs/design/unified_kernel_registry.md`
> updated.

### Why these choices

> - **Why post-load OpSlot binding instead of pre-load patching.** Keeping
>   `apply_veomni_fused_moe_patch` as the only kernel-pointer source meant
>   `auto.py` had to peek at the modeling module before `loader.load_model`
>   ran, which forced the multi-paragraph "find-an-OpSlot first, otherwise
>   pre-patch" preamble. Moving binding to `_bind_veomni_ops` (which already
>   exists for the v5 path) collapses both branches into one — the modeling
>   module is guaranteed importable by the time the loader has produced a
>   model instance, and the slot lookup naturally finds whatever the v4
>   patch entry point or the v5 patchgen put there.
>
> - **Why mirror the OpSlot onto `hf_<model>` for qwen3_moe and deepseek_v3.**
>   `_bind_veomni_ops` walks `sys.modules[model.__class__.__module__]`. On
>   transformers v4 the resolved class still lives in the upstream HF
>   module even though VeOmni has monkey-patched its `SparseMoeBlock` to a
>   `Patch*` class. Extending the bind walk to also iterate sibling VeOmni
>   modules would couple it to module-discovery heuristics; a one-line
>   `hf_<model>.veomni_moe_experts_forward = veomni_moe_experts_forward`
>   from the patch entry point is an explicit, per-model opt-in.
>
> - **Why collapse qwen3_omni_moe's eager `nn.ModuleList` path.** Keeping
>   two storage layouts (per-expert `nn.ModuleList` for eager, stacked
>   params for fused) meant the runtime checkpoint converter had to gate on
>   `_moe_implementation`, which is exactly the flag this PR is removing.
>   Stacked-params + `F.linear` is mathematically equivalent to the
>   ModuleList loop, the converter handles HF per-expert checkpoints
>   uniformly, and we delete a whole class of "did the converter fire?"
>   bug surface.
>
> - **Why keep `apply_veomni_fused_moe_patch` even though every modeling
>   file goes through OpSlot.** Removing the global `_fused_moe_forward`
>   pointer would mean routing every fused call through the slot's adapter
>   (`veomni_moe_experts_forward(self, ...)`), which would force the four
>   v4 holdouts onto a uniform storage layout (split vs. merged
>   `gate_up_proj`, transposed vs. native). That's a larger reshaping than
>   this PR's scope; filed as future work.
>
> - **Why default to `chunk_loss` instead of `eager`.** The chunked path
>   keeps activation memory bounded for large vocab and long sequences
>   without requiring a fused-kernel install (Liger). It's already what
>   every NPU run uses, and the kernel itself is pure PyTorch — one
>   hardware-agnostic default beats two device-conditional ones.
>
> - **Why preserve `moe_implementation: "fused"` after #678 split it.**
>   Pre-#678 YAMLs in production would otherwise fail at config parse on
>   upgrade. The deprecation warning gives users a window to migrate at
>   their own pace; the alias will be removed in a follow-up once internal
>   users have moved over.

### Checklist Before Submitting

- [x] Read the [Contribute Guide](https://github.com/ByteDance-Seed/VeOmni/blob/main/CONTRIBUTING.md)
- [x] Applied pre-commit checks (`make quality` clean, `make patchgen` clean)
- [x] Added/updated documentation (`transformers_v5_moe_weight_loading.md`, `kernel_selection.md`, `unified_kernel_registry.md`)
- [x] Added tests to CI workflow — covered by existing
      `tests/models/test_models_patch.py`,
      `tests/models/test_moe_v4_runtime_merge.py`,
      `tests/e2e/test_e2e_parallel.py`, and `tests/ops/test_moe_hw_gate.py`.
      No new test file needed.

### CODE REVIEW

File: veomni/models/auto.py [hunk 1]
```diff
@@ -113,23 +82,26 @@ def _bind_veomni_ops(modeling_module, ops_config: OpsImplementationConfig) -> bo
                     if ops_config.moe_implementation == "eager"
                     else ops_config.moe_implementation.removeprefix("fused_")
```
[the remark this answers is not in the document]
Comment (TimYangst) [reply veomni/models/auto.py:83]: It is intended behavior.

File: veomni/models/auto.py [hunk 2]
```diff
@@ -113,23 +82,26 @@ def _bind_veomni_ops(modeling_module, ops_config: OpsImplementationConfig) -> bo
...
+                if impl_name != "eager":
+                    moe_experts_kernel = impl_name
             else:
                 impl_name = getattr(ops_config, f"{obj.op_name}_implementation", "eager")
             obj.bind(impl_name)
             logger.info_rank0(f"OpSlot '{name}' bound to '{impl_name}' -> {obj}")
             found = True
-    return found
 
+    # When a model owns a ``moe_experts`` OpSlot, we install the global
```
Comment (Luosuu) [inline veomni/models/auto.py]: do not quite understand

**Review by Luosuu — COMMENTED** [review]
I think overall I am confused by the motivation now after so many kernel op registration PRs. Could you please clarify what is breaking now and what is the motivation for this PR?

**Comment by TimYangst** [comment]
@gemini-code-assist please help to review the PR

File: veomni/models/auto.py (hunk 2 above)
Comment (TimYangst) [reply veomni/models/auto.py]: Updated the comment.

File: tasks/deprecated_task/train_qwen_vl.py [hunk 3]
```diff
@@ -137,7 +137,6 @@ def main():
         config_path=args.model.config_path,
         weights_path=args.model.model_path,
         init_device=args.train.init_device,
-        moe_implementation=args.model.ops_implementation.moe_implementation,
         attn_implementation=args.model.ops_implementation.attn_implementation,
         encoder_data_balance=args.model.encoder_data_balance,
         encoder_data_balance_sorting_algo=args.model.encoder_data_balance_sorting_algo,
```
[the remark this answers is not in the document]
Comment (TimYangst) [reply tasks/deprecated_task/train_qwen_vl.py:141]: Fixed

File: tasks/deprecated_task/train_torch.py [hunk 4]
```diff
@@ -138,7 +138,6 @@ def main():
         weights_path=args.model.model_path,
         torch_dtype="float32" if args.train.accelerator.fsdp_config.mixed_precision.enable else "bfloat16",
         attn_implementation=args.model.ops_implementation.attn_implementation,
-        moe_implementation=args.model.ops_implementation.moe_implementation,
         init_device=args.train.init_device,
```
[the remark this answers is not in the document]
Comment (TimYangst) [reply tasks/deprecated_task/train_torch.py:140]: Fixed

File: veomni/ops/kernel_registry.py [hunk 5]
```diff
@@ -51,7 +51,12 @@ def is_satisfied(self) -> bool:
             # torch_npu package AND an actual NPU device (unlike a bare import
             # check, which passes on dev boxes that merely have the library).
             return IS_NPU_AVAILABLE
-        raise ValueError(f"Unknown device_type: {self.device_type!r} (expected 'gpu' or 'npu')")
+        if self.device_type == "any":
+            # Hardware-agnostic kernel (pure PyTorch). Used by chunk_loss
+            # which composes F.linear + eager_cross_entropy and runs on both
+            # CUDA and Ascend NPU without device-specific calls.
+            return IS_CUDA_AVAILABLE or IS_NPU_AVAILABLE
+        raise ValueError(f"Unknown device_type: {self.device_type!r} (expected 'gpu' | 'npu' | 'any')")
```
[the remark this answers is not in the document]
Comment (TimYangst) [reply veomni/ops/kernel_registry.py:61]: updated

File: veomni/models/transformers/qwen3_omni_moe/modeling_qwen3_omni_moe.py [hunk 6]
```diff
@@ -690,54 +699,78 @@ def _deepstack_process(
...
+            if expert_idx == self.num_experts:
+                continue
+            top_k_pos, token_idx = torch.where(expert_mask[expert_idx])
+            current_state = hidden_states[token_idx]
+            gate = F.linear(current_state, self.gate_proj[expert_idx])
+            up = F.linear(current_state, self.up_proj[expert_idx])
+            current_hidden_states = self.act_fn(gate) * up
+            current_hidden_states = F.linear(current_hidden_states, self.down_proj[expert_idx])
+            current_hidden_states = current_hidden_states * routing_weights[token_idx, top_k_pos, None]
+            final_hidden_states.index_add_(0, token_idx, current_hidden_states.to(final_hidden_states.dtype))
```
[the remark this answers is not in the document]
Comment (TimYangst) [reply veomni/models/transformers/qwen3_omni_moe/modeling_qwen3_omni_moe.py:771]: False positive

**Review by Luosuu — APPROVED** [review]
